src/err/mod.rs

```diff
@@ -0,0 +1,568 @@
+use crate::type_object::PyTypeObject;
+use crate::types::PyType;
+use crate::{
+    exceptions::{self, PyBaseException},
+    ffi,
+};
+use crate::{
+    AsPyPointer, FromPyPointer, IntoPy, Py, PyAny, PyNativeType, PyObject, Python,
+    ToBorrowedObject, ToPyObject,
+};
+use libc::c_int;
+use std::borrow::Cow;
+use std::cell::UnsafeCell;
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::ptr::NonNull;
+
+mod err_state;
+mod impls;
+
+pub use err_state::PyErrArguments;
+use err_state::{boxed_args, PyErrState, PyErrStateNormalized};
+
+/// Represents a Python exception that was raised.
+pub struct PyErr {
+    // Safety: can only hand out references when in the "normalized" state. Will never change
+    // after normalization.
+    //
+    // The state is temporarily removed from the PyErr during normalization, to avoid
+    // concurrent modifications.
+    state: UnsafeCell<Option<PyErrState>>,
+}
+
+unsafe impl Send for PyErr {}
+unsafe impl Sync for PyErr {}
+
+/// Represents the result of a Python call.
+pub type PyResult<T> = Result<T, PyErr>;
+
+/// Marker type that indicates an error while downcasting
+#[derive(Debug)]
+pub struct PyDowncastError<'a> {
+    from: &'a PyAny,
+    to: Cow<'static, str>,
+}
+
+impl<'a> PyDowncastError<'a> {
+    pub fn new(from: &'a PyAny, to: impl Into<Cow<'static, str>>) -> Self {
+        PyDowncastError {
+            from,
+            to: to.into(),
+        }
+    }
+}
+
+impl PyErr {
+    /// Creates a new PyErr of type `T`.
+    ///
+    /// `value` can be:
+    /// * a tuple: the exception instance will be created using Python `T(*tuple)`
+    /// * any other value: the exception instance will be created using Python `T(value)`
+    ///
+    /// Note: if `value` is not `Send` or `Sync`, consider using `PyErr::from_instance` instead.
+    ///
+    /// Panics if `T` is not a Python class derived from `BaseException`.
+    ///
+    /// Example:
+    /// ```ignore
+    /// return Err(PyErr::new::<exceptions::PyTypeError, _>("Error message"));
+    /// ```
+    ///
+    /// In most cases, you can use a concrete exception's constructor instead, which is equivalent:
+    /// ```ignore
+    /// return Err(exceptions::PyTypeError::new_err("Error message"));
+    /// ```
+    pub fn new<T, A>(args: A) -> PyErr
+    where
+        T: PyTypeObject,
+        A: PyErrArguments + Send + Sync + 'static,
+    {
+        Python::with_gil(|py| PyErr::from_type(T::type_object(py), args))
+    }
+
+    /// Constructs a new error, with the usual lazy initialization of Python exceptions.
+    ///
+    /// `exc` is the exception type; usually one of the standard exceptions
+    /// like `exceptions::PyRuntimeError`.
+    /// `args` is the a tuple of arguments to pass to the exception constructor.
+    pub fn from_type<A>(ty: &PyType, args: A) -> PyErr
+    where
+        A: PyErrArguments + Send + Sync + 'static,
+    {
+        assert_ne!(unsafe { ffi::PyExceptionClass_Check(ty.as_ptr()) }, 0);
```

Comment:
I'm not sure whether I like this kind of assertion here, on the other hand making the error-creation fallible seems weird...

At least the requirements for `ty` should be documented.

Reply:
I just saw that the constructor below handles this case by returning a `PyTypeError`, perhaps this is also the right solution here?

Reply:
👍 refactored to use the same `PyTypeError`.